worker/periodicworker_test.go

```diff
@@ -45,6 +46,43 @@ func (s *periodicWorkerSuite) TestWait(c *gc.C) {
 	}
 }
 
+func (s *periodicWorkerSuite) TestWaitWithJitter(c *gc.C) {
+	funcHasRun := make(chan struct{}, 1)
+	doWork := func(_ <-chan struct{}) error {
+		funcHasRun <- struct{}{}
+		return nil
+	}
+
+	w := NewPeriodicWorker(doWork, defaultPeriod, NewTimer, Jitter(0.2))
+	defer func() { c.Assert(worker.Stop(w), gc.Equals, nil) }()
+
+	select {
+	case <-funcHasRun:
+	case <-time.After(testing.ShortWait):
+		c.Fatalf("The doWork function should have been called by now")
+	}
+
+	t := time.Now()
+	var actualPeriod time.Duration
+	select {
+	case <-funcHasRun:
+		actualPeriod = time.Since(t)
+	case <-time.After(2 * defaultPeriod):
+		c.Fatalf("The doWork function should have been called by now")
+	}
+	c.Assert(float64(actualPeriod)/float64(defaultPeriod) <= 1.2, jc.IsTrue)
+	c.Assert(float64(actualPeriod)/float64(defaultPeriod) >= 0.8, jc.IsTrue)
```

Comment:
I'm concerned that this is going to end up being flaky. I've seen Go's timers be quite inaccurate in our test suites, going off both much later *and* earlier than requested, especially on busy or underprovisioned hosts. Notice how none of the other tests specifically check the period. The way that the worker is currently structured, this can't be reliably tested.

Possible solutions:
* set wide margins the allowed period - but I think they'll actually need to be so wide that they'll be meaningless
* use an internal test which ensures that `w.jitter` is set and as the required effect (by factoring out `nextPeriod` as suggested)
* rework the worker in terms of an injected Clock like we do elsewhere so that tests aren't involved with wall time (the best solution but also the most time consuming)